src/XMakeBuildEngine/Evaluation/MSBuildExtensionsPathHelper.cs

```diff
@@ -0,0 +1,164 @@
+            switch(refKind) {
+                case MSBuildExtensionsPathReferenceKind.Default:
+                    return "MSBuildExtensionsPath";
+                case MSBuildExtensionsPathReferenceKind.Path32:
+                    return "MSBuildExtensionsPath32";
+                case MSBuildExtensionsPathReferenceKind.Path64:
+                    return "MSBuildExtensionsPath64";
+                case MSBuildExtensionsPathReferenceKind.None:
+                default:
+                    return null;
+            }
+        }
+
+        ///<summary>
+        /// Helper method for resetting the lists for testing, since we need to run the
+        /// tests with different values for the *SearchPaths
+        ///</summary>
+        private static void ResetPathsListsForTesting()
+        {
+            s_pathsLists = new Lazy<ExtensionsPathsLists>(true);
+        }
+
+        class ExtensionsPathsLists
+        {
+            /// <summary>
+            /// Cached list of extension paths to be used as search paths for $(MSBuildExtensionsPath)
+            /// </summary>
+            public readonly List<string> Default;
+
+            /// <summary>
+            /// Cached list of extension paths to be used as search paths for $(MSBuildExtensionsPath32)
+            /// </summary>
+            public readonly List<string> Path32;
+
+            /// <summary>
+            /// Cached list of extension paths to be used as search paths for $(MSBuildExtensionsPath64)
+            /// </summary>
+            public readonly List<string> Path64;
+
+            public ExtensionsPathsLists()
+            {
+                Default = ComputeFallbackExtensionsPathsFor(MSBuildExtensionsPathReferenceKind.Default);
+                Path32 = ComputeFallbackExtensionsPathsFor(MSBuildExtensionsPathReferenceKind.Path32);
+                Path64 = ComputeFallbackExtensionsPathsFor(MSBuildExtensionsPathReferenceKind.Path64);
+            }
+
+            private static List<string> ComputeFallbackExtensionsPathsFor(MSBuildExtensionsPathReferenceKind refKind)
+            {
+                var extnPaths = new List<string>();
+                var refKindAsString = ReferenceKindToString(refKind);
+                var envVar = Environment.GetEnvironmentVariable(refKindAsString + "SearchPaths");
+
+                var allPaths = String.Join(Path.PathSeparator.ToString(),
+                                            //FIXME: IsOSX returns false :/
+                                            NativeMethodsShared.IsUnix ? s_osxExternalXBuildDir : String.Empty,
+                                            envVar ?? String.Empty);
+
+                var pathsTable = new Dictionary<string, string>();
+                foreach(var extnPath in allPaths.Split(new char[]{Path.PathSeparator}, StringSplitOptions.RemoveEmptyEntries)) {
+                    if(pathsTable.ContainsKey(extnPath))
+                        continue;
+
+                    if(!Directory.Exists(extnPath))
```

Comment:
Why not let callers check for existence? What if the caller code wants to do some funky lazy creation of the paths?